interp/src/primitives/combinational.rs

```diff
@@ -67,37 +74,187 @@ comb_primitive!(StdNot[WIDTH](r#in: WIDTH) -> (out: WIDTH) {
+        )
+        .into_iter()
+        .next()
+        .map(|(_, v)| v)
+        .unwrap()
+        .unwrap_imm();
+
+    //now do addition. maybe better to use the adder and unwrap the OutputValue?
+    let a_iter = left.vec.iter().by_ref();
+    let b_iter = new_right.vec.iter().by_ref();
+    let mut c_in = false;
+    let mut sum = BitVec::new();
+    for (ai, bi) in a_iter.zip(b_iter) {
+        sum.push(
+            c_in & !ai & !bi
+                || bi & !c_in & !ai
+                || ai & !c_in & !bi
+                || ai & bi & c_in,
+        );
+        c_in = bi & c_in || ai & c_in || ai & bi || ai & c_in & bi;
+    }
+    //actually ok if there is overflow from 2sc subtraction (?)
+    //have to check if this is ok behavior
+    return Value { vec: sum }.into();
+    todo!()
+});
+comb_primitive!(StdDivPipe[WIDTH](left: WIDTH, right: WIDTH) -> (out: WIDTH) {
+    todo!()
+});
+// ===================== Signed binary operations ======================
+comb_primitive!(StdSmultPipe[WIDTH](left: WIDTH, right: WIDTH) -> (out: WIDTH) {
+    todo!()
+});
+comb_primitive!(StdSdivPipe[WIDTH](left: WIDTH, right: WIDTH) -> (out: WIDTH) {
+    todo!()
+});
+// ===================== Unsigned FP binary operations ======================
+comb_primitive!(StdFpAdd[WIDTH](left: WIDTH, right: WIDTH) -> (out: WIDTH) {
+    todo!()
+});
+comb_primitive!(StdFpSub[WIDTH](left: WIDTH, right: WIDTH) -> (out: WIDTH) {
+    todo!()
+});
+comb_primitive!(StdFpMultPipe[WIDTH](left: WIDTH, right: WIDTH) -> (out: WIDTH) {
+    todo!()
+});
+comb_primitive!(StdFpDivPipe[WIDTH](left: WIDTH, right: WIDTH) -> (out: WIDTH) {
+    todo!()
+});
+
+// ===================== Signed FP binary operations ======================
+comb_primitive!(StdFpSadd[WIDTH](left: WIDTH, right: WIDTH) -> (out: WIDTH) {
+    todo!()
+});
+comb_primitive!(StdFpSsub[WIDTH](left: WIDTH, right: WIDTH) -> (out: WIDTH) {
+    todo!()
+});
+comb_primitive!(StdFpSmultPipe[WIDTH](left: WIDTH, right: WIDTH) -> (out: WIDTH) {
+    todo!()
+});
+comb_primitive!(StdFpSdivPipe[WIDTH](left: WIDTH, right: WIDTH) -> (out: WIDTH) {
+    todo!()
 });
 
 // ===================== Shift Operations ======================
 comb_primitive!(StdLsh[WIDTH](left: WIDTH, right: WIDTH) -> (out: WIDTH) {
-    let mut tr = left.vec.clone();
-    tr.shift_right(right.as_u64() as usize);
-    Value { vec: tr }.into()
+    //to avoid the casting overflow,
+        //we know that [left], [right], and [self]
+        //are capped at bitwidths as large as largest u64 (2^64 - 1 = 1.84 * 10^19 ...)
+        //so, if [self] has a width greater than 64,
+        //and the 65th index is a 1, we just automatically return a 0 of the
+        //appropriate bitwidth!
+
+        if WIDTH > 64 {
+            //check if right is greater than or equal to  2 ^ 64
+            let r_vec = &right.vec;
+            for (index, bit) in r_vec.iter().by_ref().enumerate() {
+                if (index >= 64) & *bit {
```

Comment:
You might consider using something like [`.skip(64)`](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.skip) instead of iterating over everything and ignoring the first 64?